src/methods/hash.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { BabelBinaryOperator, BabelUnaryOperator } from "babel-types";
+import invariant from "../invariant.js";
+
+interface Hashable { getHash(): number, mightNotBeNumber(): boolean }
+
+export function hashBinary<T: Hashable>(op: BabelBinaryOperator, x: T, y: T): [number, Array<T>] {
+  let xHash = x.getHash();
+  let yHash = y.getHash();
+  if (yHash < xHash) {
+    switch (op) {
+      case "+":
+      case "*":
+      case "==":
+      case "===":
+      case "!=":
+      case "!==":
+        if (op !== "+" || !(x.mightNotBeNumber() || y.mightNotBeNumber())) {
+          [x, y] = [y, x];
+          [xHash, yHash] = [yHash, xHash];
+        }
+        break;
+      default:
+        break;
+    }
+  }
+  let hash = (((hashString(op) * 13) ^ xHash) * 13) ^ yHash;
```

Comment:
It took me a moment to convince myself that nothing bad is ever happening, even if a `NaN` hash from a number value enters the picture.